xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java

```diff
@@ -342,6 +342,129 @@ protected void initializeCoreSchema() throws SolrException
+                // Add the missing type
+                this.solrSchemaUtils.setFieldType(targetCore, entry.getValue(), true);
+            }
+        }
+
+        commit(targetCore);
+    }
+
+    private void migrateFields(XWikiSolrCore sourceCore, XWikiSolrCore targetCore) throws SolrException
+    {
+        migrateFields(sourceCore, targetCore, true);
+        migrateFields(sourceCore, targetCore, false);
+
+        commit(targetCore);
+    }
+
+    private void migrateFields(XWikiSolrCore sourceCore, XWikiSolrCore targetCore, boolean dynamic) throws SolrException
+    {
+        Map<String, Map<String, Object>> sourceFields =
+            dynamic ? this.solrSchemaUtils.getDynamicFields(this.core, false)
+                : this.solrSchemaUtils.getFields(sourceCore, false);
+        Map<String, Map<String, Object>> targetFields =
+            dynamic ? this.solrSchemaUtils.getDynamicFields(this.core, false)
+                : this.solrSchemaUtils.getFields(targetCore, false);
+        for (Map.Entry<String, Map<String, Object>> entry : sourceFields.entrySet()) {
+            if (!targetFields.containsKey(entry.getKey())) {
+                // Add the missing type
+                this.solrSchemaUtils.setField(targetCore, entry.getValue(), dynamic, true);
+            }
+        }
+    }
+
+    private void migrateData(XWikiSolrCore sourceCore, XWikiSolrCore targetCore) throws SolrException
+    {
+        int size = 0;
+        do {
+            SolrQuery solrQuery = new SolrQuery();
+            solrQuery.setRows(MIGRATION_BATCH_ROWS);
+
+            QueryResponse response;
+            try {
+                response = sourceCore.getClient().query(solrQuery);
+            } catch (Exception e) {
+                throw new SolrException("Failed to search for entries in the source client", e);
+            }
+
+            SolrDocumentList result = response.getResults();
+            size = result.size();
+
+            if (size > 0) {
+                migrateData(response.getResults(), sourceCore, targetCore);
+            }
+        } while (size == MIGRATION_BATCH_ROWS);
+    }
+
+    private void migrateData(SolrDocumentList source, XWikiSolrCore sourceCore, XWikiSolrCore targetCore)
+        throws SolrException
+    {
+        for (SolrDocument sourceDocument : source) {
+            SolrInputDocument targetDocument = new SolrInputDocument();
+
+            // Convert the SolrDocument into a SolrInputDocument
+            migrate(sourceDocument, targetDocument);
+
+            // Save the document
+            try {
+                targetCore.getClient().add(targetDocument);
+            } catch (Exception e) {
+                throw new SolrException("Failed to save the document", e);
+            }
+
+            // Add it to the list of documents to delete
+            delete(source, sourceCore);
+        }
+
+        // Delete the documents
+        delete(source, targetCore);
+        commit(sourceCore);
```

Comment:
Why is the source but not the target committed? Wouldn't it make sense to first commit the target and then the source to ensure no data is lost?